WEB-INTEGRATION.md

```diff
@@ -0,0 +1,742 @@
+  useful to expose, since the call to `intercept()` can be seen as scheduling
+  those callbacks, and the fact that there is JS code on the call stack can be
+  seen as an implementation detail.
+
+### Action registrations
+
+These APIs register a callback or constructor to be invoked when some action
+runs. They’re also commonly used as a way to associate a newly created class
+instance with some action, such as in worklets or with custom elements.
+
+In cases where the action that causes the callback or constructor to be invoked
+originates with the browser, not being caused by any JS code, then the callback
+will have the same context as when the API that registered it was called. For
+worklets, only JS code inside the worklet counts as causing the action. In these
+cases, there is no other AsyncContext snapshot to consider using besides the one
+which was active when the API was called, so the generic WebIDL solution works.
+
+- [`navigator.mediaSession.setActionHandler()`](https://w3c.github.io/mediasession/#dom-mediasession-setactionhandler)
+  method [\[MEDIASESSION\]](https://w3c.github.io/mediasession/)
+- [`navigator.geolocation.watchPosition()`](https://w3c.github.io/geolocation/#dom-geolocation-watchposition)
+  method [\[GEOLOCATION\]](https://w3c.github.io/geolocation/)
+- [`RemotePlayback`](https://w3c.github.io/remote-playback/#dom-remoteplayback):
+  [`watchAvailability()`](https://w3c.github.io/remote-playback/#dom-remoteplayback-watchavailability)
+  method [\[REMOTE-PLAYBACK\]](https://w3c.github.io/remote-playback/)
+- [`AudioWorkletGlobalScope`](https://webaudio.github.io/web-audio-api/#AudioWorkletGlobalScope):
+  [`registerProcessor()`](https://webaudio.github.io/web-audio-api/#dom-audioworkletglobalscope-registerprocessor)
+  method [\[WEBAUDIO\]](https://webaudio.github.io/web-audio-api/)
+- [`PaintWorkletGlobalScope`](https://drafts.css-houdini.org/css-paint-api-1/#paintworkletglobalscope):
+  [`registerPaint()`](https://drafts.css-houdini.org/css-paint-api-1/#dom-paintworkletglobalscope-registerpaint)
+  method [\[CSS-PAINT-API\]](https://drafts.css-houdini.org/css-paint-api-1/)
+
+In the case of custom elements, the actions that invoke the custom element
+methods are usually triggered from JS code calling a web API, e.g. removing a
+custom element with `el.remove()` will cause its `disconnectedCallback` method
+to be run. But in some cases custom element reactions can also run as a response
+to a user action, such as removing a custom element inside a text editing
+context. To avoid having a difference between these cases, we propose using the
+context in which the custom element was defined with
+[`customElements.define()`](https://html.spec.whatwg.org/multipage/custom-elements.html#dom-customelementregistry-define)
+[\[HTML\]](https://html.spec.whatwg.org/multipage/). In the future it might be
+possible to add a way to get the snapshot in which el.remove() was invoked, if
+there are use cases.
```

Comment:
When discussing AsyncContext with @yoavweiss, he also mentioned that custom element reactions are a case where you'd want to go with call-time rather than registration-time. But it's not clear to me why this would be the case and whether it's unique to custom element reactions. And if WebIDL automatically takes care of storing and running the registration snapshot, it might not be easy spec-wise or implementation-wise to go with call-time.

The thing that convinced me that custom element reactions should use the registration snapshot, however, is the fact that they don't always have a call snapshot. If you have a custom element inside a `contentEditable` element, and you delete it when editing, there's no call context for the `disconnectedCallback` reaction. So that would break one of the design goals we set for the web integration, which was that the context in which a callback is called from a web API would not be either empty or non-empty depending on factors unrelated to the web API.

Reply:
I discussed this with @rictic and he came to the same conclusions that registration snapshot would be basically useless for custom element lifecycle events.  In the case where there _is_ a workable programmatic causal snapshot, I think it needs to be the default in order to not break ~every use case for AsyncContext when used with custom elements.  We can discuss elsewhere whether the fallback should be a null/top-level snapshot, or else the registration snapshot.

Reply:
Call-time semantics for custom elements give you the context that connected/disconnected the element, which is what you want (at least for Task Attribution in the context of soft navigations).

The same could be true for e.g. PerformanceObsrevers. E.g. for user timing, you'd want the context of the code that added a mark, rather than the generic RUM code that registered the handler.

Reply:
^^ @shaseley @mmocny